metapool/sample_sheet.py

```diff
@@ -1488,42 +1524,99 @@ def __init__(self, path=None):
+    # value to include all columns, name them with integers, and later
+    # truncate all columns that are entirely empty.
+    df.dropna(how='all', axis=1, inplace=True)
 
-    sheet = AmpliconSampleSheet(sample_sheet_path)
-    if sheet.validate_and_scrub_sample_sheet(echo_msgs=False):
-        return sheet
+    # remove all whitespace rows (drop all rows that are entirely empty)
+    df.dropna(how='all', axis=0, inplace=True)
 
-    sheet = MetagenomicSampleSheetv102(sample_sheet_path)
-    if sheet.validate_and_scrub_sample_sheet(echo_msgs=False):
-        return sheet
+    # remove all comments rows, whether they are at the top of the file (no
+    # longer supported, technically), or not.
+    comment_rows = df.index[df[0].str.startswith("#")].tolist()
+    df = df.drop(index=comment_rows)
+    # reset the index to make it easier to post-process.
+    df.reset_index(inplace=True, drop=True)
 
-    sheet = MetagenomicSampleSheetv101(sample_sheet_path)
-    if sheet.validate_and_scrub_sample_sheet(echo_msgs=False):
-        return sheet
+    # for simplicity's sake, assume the first row marks the [Header]
+    # column and raise an Error if not. By convention it should be, once
+    # legacy comments and whitespace rows are removed.
+    if df[0][0] != '[Header]':
+        raise ValueError("Top section is not [Header]")
 
-    sheet = MetagenomicSampleSheetv100(sample_sheet_path)
-    if sheet.validate_and_scrub_sample_sheet(echo_msgs=False):
-        return sheet
+    # identify the beginning of the following section and remove everything
+    # from the start of the first section on down. Remove the now redundant
+    # [Header] from the top row as well.
+    next_section_start = df.index[df[0].str.startswith("[")].tolist()[1]
+    df = df.iloc[1:next_section_start]
 
-    sheet = MetagenomicSampleSheetv90(sample_sheet_path)
-    if sheet.validate_and_scrub_sample_sheet(echo_msgs=False):
-        return sheet
+    # lastly, trim off the additional all-empty columns that are now
+    # present after the removal of the other sections.
+    df.dropna(how='all', axis=1, inplace=True)
 
-    sheet = MetatranscriptomicSampleSheetv10(sample_sheet_path)
-    if sheet.validate_and_scrub_sample_sheet(echo_msgs=False):
-        return sheet
+    # set the index to the attributes column of the sample-sheet, replacing
+    # the numeric index which now isn't needed. The dataframe will now just
+    # contain the index column and a single column named 1.
+    df.set_index(0, inplace=True)
+
+    # return the value of key '1'. This will return an immediately
+    # recognizable dictionary of key/value pairs.
+    results = df.to_dict()[1]
+
+    # conversion to dict causes SheetVersion to be wrapped in single ticks.
+    # e.g.: "'100'". These should be removed if present.
+    if 'SheetVersion' in results:
+        results['SheetVersion'] = results['SheetVersion'].replace("'", "")
+
+    return results
+
+
+def load_sample_sheet(sample_sheet_path):
```

Comment:
Should this be a member method of a sample sheet base class?

Reply:
Punt @charles-cowart

Reply:
Historically it's been a function and it's used in probably every sample-sheet so it makes sense to keep it the way it is. Moreover, associating it with the KLSampleSheet() class means the base class would have to have knowledge of all its children which breaks separation of concerns.

Another reason is that a user can currently import load_sample_sheet() without having to import any of the growing number of SampleSheet() classes or remember their names. We could make it a static method of KLSampleSheet(), which we could then export. The user would then have to import KLSampleSheet and then use the static method to load the data into the right child class.

However I think this would make things potentially more confusing because users may try to just go back to doing sheet = KLSampleSheet(sheet_path) and this no longer works because we don't allow the base class to be instantiated anymore. We also want to prevent users from manually trying to populate a KLSampleSheet() object which was done in the early test code because this sidesteps validation. In the end there's no benefit gained from making it a member method.

Reply:
Ah, okay ya so this is essentially a factory pattern. Moving most of logic in this method into the base class likely would simplify and further generalize the factory

Reply:
Yes, it is. I'm not supportive of changing the current interface out from under the notebook users w/out their support. @AmandaBirmingham do you think this change would be okay w/the wetlab? We'd also have to make a pass over the notebooks to update how sample-sheet files are loaded.